### ISSUE TITLE
Why are the mean value of model weights all 0?

### ISSUE THREAD

**Issue by YoucanBaby** [tencent-ailab/IP-Adapter#417]
Dear author,

Thanks a lot for your great project.

Why is the mean of model weights all 0? 
Did you use any training tricks?

<img width="465" alt="1723528894801" src="https://github.com/user-attachments/assets/c0839d9b-174e-41c2-981e-37d2484d29cb">

Best regards.

**Comment by ssshamiii** [comment]
Hello, I am also curious about it. Have you got any clue about this?

**Comment by YoucanBaby** [comment]
> Hello, I am also curious about it. Have you got any clue about this?

Sorry, no clues so far.

**Comment by jxl0131** [comment]
This is probably because there is a torch.nn.LayerNorm before the to_k_ip linear layer. Layernorm, which normalizes the input of a linear layer to a mean of 0. Let the input of this linear layer be x = [ x 1, x 2, ... XN ] , the output is y = [ y 1, y 2, ... y M]. Layernorm makes mean (x) = 0. During the gradient back propagation, y1 = x1x11 + x2w21 + ... xnwn1. The gradient of the layers after y 1 is represented by D, the sum of updates to the weight w11 to wn1 is sum (x1D + x2D + ... xnD) = mean(x) N D= 0. **This means that the sum of the to_k_ip weights during training is the same as at the beginning of the training initialization**, and i guess the sum of the weights at initialization is zero.

**Comment by ssshamiii** [comment]
> This is probably because there is a torch.nn.LayerNorm before the to_k_ip linear layer. Layernorm, which normalizes the input of a linear layer to a mean of 0. Let the input of this linear layer be x = [ x 1, x 2, ... XN ] , the output is y = [ y 1, y 2, ... y M]. Layernorm makes mean (x) = 0. During the gradient back propagation, y1 = x1x11 + x2w21 + ... xnwn1. The gradient of the layers after y 1 is represented by D, the sum of updates to the weight w11 to wn1 is sum (x1D + x2D + ... xnD) = mean(x) N D= 0. **This means that the sum of the to_k_ip weights during training is the same as at the beginning of the training initialization**, and i guess the sum of the weights at initialization is zero.

Your analysis makes sense, but we need to consider the impact of the **scale ($\gamma$)** and **shift ($\beta$)** parameters in `LayerNorm`. These parameters alter the distribution of the normalized input, causing the mean of the input to no longer be strictly zero.

Specifically:
1. The output of `LayerNorm` is $x' = \gamma \cdot \text{Norm}(x) + \beta $, where $\beta$ shifts the mean of $x'$ away from zero.
2. During backpropagation, weight updates depend on the values of $x'$. Due to the presence of $\beta$, $\sum x'$ is generally non-zero, meaning the weights receive non-zero updates.
3. This implies that the weights in the `to_k_ip` layer will change during training, and their sum will not remain constant at the initialized value.

Therefore, the $\beta$ parameter in `LayerNorm` breaks the assumption of zero-mean input, resulting in non-zero weight updates. As a result, the sum of the weights will dynamically change during training.
